cli: Improve -getinfo return format

klementtan opened the pull request:
# Overview
**Description** This PR changes the return format of `-getinfo`

**Rationale**  
- make `-getinfo` more user-friendly
- uses less vertical screen space.

Fixes: Issue 17314(Not linking issue to prevent it from closing)
> Alternative, more human-friendly, format besides JSON? Colors, progress bars

### Return Format
```bash
// Data from getblockchaininfo
Chain: [getblockchaininfo.chain] 
Blocks: [getblockchaininfo.blocks]
Headers: [getblockchaininfo.headers]
Verification progress: [getblockchaininfo.verificationprogress]
Difficulty: [getblockchaininfo.difficulty]

# Data from getnetworkinfo
Network: in [getnetworkinfo.connections_in], out [getnetworkinfo.connections_out], total [getnetworkinfo.connections]
Version: [getnetworkinfo.version]
Time offset (s): [getnetworkinfo.timeoffset]
Proxy: [getnetworkinfo.proxy] // "N/A" if no proxy
Min tx relay fee rate (BTC/kvB): [getnetworkinfo.relayfee]

# Data from getwalletinfo. Will only be present when a single wallet is loaded
Wallet: [getnetworkinfo.walletname] // "" if walletname is empty(default wallet)
Keypool size: [getnetworkinfo.keypoolsize]
Unlocked until: [getnetworkinfo.unlocked_until]
Transaction fee rate (-paytxfee) (BTC/kvB): [getnetworkinfo.paytxfee]

# Data from getbalances. Will only be present when a single wallet is loaded
Balance: [getbalances.mine.trusted]

# Data from listwallets then getbalances for each wallet. Will only be present for multiple wallets loaded
Balances
[getbalances.mine.trusted] [listwallets[i]] // Right justify on balance

Warnings: [getnetworkinfo.warnings]
```
#### Coloring
The following lines would be colored to better show the differences between the various `-getinfo` components. However, this will not apply for `WIN32`(ANSI code not supported) and users that connect the `stdout` to a non-terminal process.
- `Chain: ...`:  BLUE(`\x1B[34m`)
- `Network: ...`: GREEN(`\x1B[32m`)
- `Wallet: ...`: MAGENTA(`\x1B[35m`)
- `Balance: ...` CYAN(`\x1B[36m`)
- `Balances: ...` CYAN(`\x1B[36m`)
- `Warnings: ...` Yellow(`\x1B[33m`)


### Screenshots

*No wallet loaded:*
![image](https://user-images.githubusercontent.com/49265907/121794631-94b62080-cc3c-11eb-8627-d0d1c25f0878.png)

*Single wallet loaded*
![image](https://user-images.githubusercontent.com/49265907/121794619-6df7ea00-cc3c-11eb-8420-d0c18236b188.png)

*Multi wallet loaded*
![image](https://user-images.githubusercontent.com/49265907/121794626-810aba00-cc3c-11eb-8cd7-c15ede1918ac.png)

# Reviewing Notes

## Testing Scenarios

**1. No wallet loaded**
- When there no wallets are loaded(Unload wallet with `./src/bitcoin-cli unloadwallet "YOUR_WALLETNAME"`
-  Test with `./src/bitcoin-cli -getinfo`.
- Should only display `chain` and `network` segment

**2. Single wallet loaded**
- When only a single wallet loaded or `-rpcwallet` is set(Load wallet with `./src/bitcoin-cli loadwallet "YOUR_WALLETNAME"`)
- Test with `./src/bitcoin-cli -rpcwallet="YOUR_WALLETNAME" -getinfo`(Load wallet with `./src/bitcoin-cli loadwallet "YOUR_WALLETNAME"`)
- Should only display `chain`, `network`, `wallet` and `balance` segment

**3. Multiple wallet loaded**
- When there are multiple wallets loaded(Load wallet with `./src/bitcoin-cli loadwallet "YOUR_WALLETNAME"`)
- Test with `./src/bitcoin-cli -getinfo`
- Should only display `chain`, `network` and `balances` segment

## Implementation

**Current Flow**
1. `GetinfoRequestHandler` is instantiated
2. `ConnectAndCallRPC` is called with `GetinfoRequestHandler`. It returns `UniValue::VOBJ result`
3. If `-rpcwallet` wallet is not set, `GetWalletBalances` is called with a pointer to `result` as a parameter. It adds the balances for all the wallets to `result`

**New Flow**
1. `GetinfoRequestHandler` is instantiated
2. `ConnectAndCallRPC` is called with `GetinfoRequestHandler`. It returns `UniValue::VOBJ result`
3. If `-rpcwallet` wallet is not set, `GetWalletBalances` is called with a pointer to `result` as a parameter. It adds the balances for all the wallets to `result`
4. **New** `ParseGetInfoResult` is called with `result` as parameter. It converts results type from `UniValue::VOBJ` to `UniValue::VSTR` according to the format stated above.

klementtan committed:
cli: Implement human readable -getinfo.

jonatack commented:
Concept ACK. Quick test below on signet. Build is clean, only skimmed the code for now.

Ideas:
- perhaps try displaying the balances before the wallet names, as the balances have a fixed length
- maybe display the connections on one line `connections: in 10, out 10, total 20`...a good opportunity to save on vertical space, as getinfo was previously pruned down due to its output being considered too long
- maybe s/Blockchain/Block chain/ (see https://github.com/bitcoinops/bitcoinops.github.io/blob/master/STYLE.md for a bitcoin writing guide)
- add spaces between words like verificationprogress, relayfee, and timeoffset
- maybe display `""` for the default wallet

![Screenshot from 2021-05-03 12-10-02](https://user-images.githubusercontent.com/2415484/116864897-22433100-abf8-11eb-93e4-b7c19ef0fb4b.png)

klementtan commented:
@jonatack 
Updated the PR according to your comments:
1. Swap position of wallet and balance and 
2. Prefix balance  value with ₿ emoji to remove any ambiguity of what it represents 
3.  display the connections on one line 
4. changed `Blockchain` to `Block chain`
5. added space to `verification progress`, `realy fee`, `time offset` and `keypool size`(Should I also change `paytxfee` to `pay tx fee`?)
6. display `""` for default wallet 

Multiple wallet loaded
![image](https://user-images.githubusercontent.com/49265907/116883901-b1803100-ac58-11eb-8b64-3a916ae1086b.png)
 
Single wallet loaded
![image](https://user-images.githubusercontent.com/49265907/116884081-e8eedd80-ac58-11eb-898a-9fa41b66be1f.png)

laanwj commented:
I think this is nice. It looks better, is more user friendly, and at the same time discourages parsing the output in scripts and such (which is good because it is not intended as a stable interface).

laanwj commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +868,62 @@ static void GetWalletBalances(UniValue& result)
+        return;
+    }
+
+    std::string RESET = "\033[0m";
```
Are there still issues on windows with colorized text, or can basic ANSI color support be assumed all over the board now?

klementtan replied:
hmmm I do not own a Windows computer and cannot test if ANSI color support works. 

> (if not, you might want to put this in #ifndef WIN32, leave the strings empty if not—we do this for some python scripts)
Am I right to say you are referring to something like this? 
```
    std::string RESET = "";
    std::string GREEN = "";
    std::string BLUE = "";
    std::string YELLOW = "";
    std::string MAGENTA = "";
    std::string CYAN = "";
#ifndef WIN32
    RESET = "\033[0m";
    GREEN = "\x1B[32m";
    BLUE = "\x1B[34m";
    YELLOW = "\x1b[33m";
    MAGENTA = "\x1B[35m";
    CYAN = "\x1B[36m";
#endif
```

laanwj replied:
Yes, for example. I think I would personally do, to make it symmetrical:
```c++
#ifndef WIN32
    const std::string RESET = "\033[0m";
    const std::string GREEN = "\x1B[32m";
    const std::string BLUE = "\x1B[34m";
    const std::string YELLOW = "\x1b[33m";
    const std::string MAGENTA = "\x1B[35m";
    const std::string CYAN = "\x1B[36m";
#else
    const std::string RESET = "";
    const std::string GREEN = "";
    const std::string BLUE = "";
    const std::string YELLOW = "";
    const std::string MAGENTA = "";
    const std::string CYAN = "";
#endif
```

klementtan replied:
@laanwj Yeah you're right that looks better 👍. Updated the PR to add WIN32 check.

klementtan commented:
@laanwj Roger that. Squashed all the commits to a single commit

laanwj commented:
Thanks!

> Not sure why the last CI task is failing tho. Anyone have a clue?

Seems to be solved now. We have transient issues with the CI quite often, unfortunately.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +868,71 @@ static void GetWalletBalances(UniValue& result)
     result.pushKV("balances", balances);
 }
 
+static void ParseGetInfoResult(UniValue& result)
```
Could add Doxygen documentation for this new function.

klementtan replied:
Added Doxygen documentation for `ParseGetInfoResult`

klementtan replied:
👍 updated it according to your suggestion.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +868,71 @@ static void GetWalletBalances(UniValue& result)
+    if (!result["balances"].isNull()) {
+        result_string += strprintf("%sBalances%s\n", CYAN, RESET);
+
+        for (std::string wallet : result["balances"].getKeys()) {
```
Can reference to const here rather than copying
 ```suggestion
        for (const std::string& wallet : result["balances"].getKeys()) {
```

klementtan replied:
Updated to use string reference instead of copying

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +868,71 @@ static void GetWalletBalances(UniValue& result)
+{
+    if (!find_value(result, "error").isNull()) {
+        return;
+    }
```
style nit, for a guard conditional at the top of the function this seems more compact to me, feel free to ignore
```diff
-    if (!find_value(result, "error").isNull()) {
-        return;
-    }
+    if (!find_value(result, "error").isNull()) return;
```

klementtan replied:
Followed your suggestion and updated to a single line guard clause

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -991,6 +1057,11 @@ static int CommandLineRPC(int argc, char *argv[])
+                if (gArgs.IsArgSet("-getinfo")) {
+                    ParseGetInfoResult(result);
+                }
+
```
nit, maybe combine the two -getinfo checks, feel free to ignore
```diff
             if (error.isNull()) {
-                if (gArgs.IsArgSet("-getinfo") && !gArgs.IsArgSet("-rpcwallet")) {
-                    GetWalletBalances(result); // fetch multiwallet balances and append to result
-                }
-
                 if (gArgs.IsArgSet("-getinfo")) {
+                    if (!gArgs.IsArgSet("-rpcwallet")) {
+                        GetWalletBalances(result); // fetch multiwallet balances and append to result
+                    }
                     ParseGetInfoResult(result);
                 }
```

klementtan replied:
Followed your suggestion and combined the checks

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -991,6 +1057,11 @@ static int CommandLineRPC(int argc, char *argv[])
                     GetWalletBalances(result); // fetch multiwallet balances and append to result
                 }
+
+                if (gArgs.IsArgSet("-getinfo")) {
```
nit, after using `if (gArgs.IsArgSet("-getinfo"))` a year ago, I learned later that this does more robust value checking of the user input:
```suggestion
                if (gArgs.GetBoolArg("-getinfo", false)) {
```

klementtan replied:
Updated to use `GetBoolArg` instead of `IsArgSet`

jonatack reviewed (commented):
Quick skim of the code, waiting on the build to test.

jonatack commented:
Approach ACK

Test:
```
Network
version: 219900
time offset: 0
proxy:  
connections: in 12, out 10, total 22
relay fee: 0.00001000

Block chain
blocks: 37466
headers: 37466
verification progress: 0.999984530335539
difficulty: 0.00287632828563724
chain: signet

Balances
₿0.20997653: default settings
₿0.00000000: disable private keys and blank
₿0.00000000: disable private keys
₿0.00000000: encrypted blank descriptor
₿0.00000000: blank
₿509.87584523: ""
```

Some questions (not sure they are good ideas!)

The Bitcoin symbols are neat but I'm not sure they are needed...or maybe insert a space between them and the balance.

Should the field names be capitalized? e.g. s/verification progress/Verification progress/

Maybe hoist "chain: signet" to the header and shorten it to chain? e.g. "Chain: signet"

jonatack commented:
![Screenshot from 2021-05-12 15-25-10](https://user-images.githubusercontent.com/2415484/117982651-c0847480-b325-11eb-95b1-c0a315ec1a8f.png)

klementtan commented:
Updated the PR to according @jonatack' comments
> Some questions (not sure they are good ideas!)  
The Bitcoin symbols are neat but I'm not sure they are needed...or maybe instead add just one symbol in parentheses after "Balances" in the header, e.g. Balances (B)...or insert a space between them and the balance?
Should the field names be capitalized? e.g. s/verification progress/Verification progress/?
Maybe hoist "chain: signet" to the header and shorten it to chain? e.g. "Chain: signet", thereby removing a line...and if so, should that section be at the top, before the Network section?
1. Capitalized all field names. s/verification progress/Verification progress/
2. Changed `unlocked_until` to `Unlocked until` and `paytxfee` to `Pay transaction fee`
3. Changed `₿0.000: wallet` to `Balances(₿) ... 0.000: wallet`
4. Combined `chain: signet` to `Chain(signet)`
5. Moved `Chain` to be the top of the response.
6. Updated `cli_get_info_string_to_dict` in test to a static method and changed string parsing method.  

Single wallet loaded
![image](https://user-images.githubusercontent.com/49265907/118020467-b053b000-b38c-11eb-8ebc-022986371822.png)

Multi wallet loaded
![image](https://user-images.githubusercontent.com/49265907/118020563-c82b3400-b38c-11eb-8c75-9afe83c6e704.png)

jonatack commented:
Test of f93b021472

![Screenshot from 2021-05-12 22-48-59](https://user-images.githubusercontent.com/2415484/118042593-cd27bd80-b363-11eb-9c4c-49c989f85b5e.png)

What do you think about
- `s/Balance(₿):/Balance (₿):/` (insert a space)
- extend the colors to the full line of `Chain: <chain>` and `Balances (₿)`
- right-justify the balances to the largest balance, prefixing spaces to the smaller balances (-netinfo in this same file does it for some of the fields using strprintf)
- maybe (not sure, might be too cluttered) hoisting the connections to the header: `Network: in 17, out 22, total 49` (single color)

jonatack commented:
Oh, and if you need some sBTC (signet coins) for testing, here's a faucet: https://signetfaucet.com

klementtan commented:
Updated [ed7feb4](https://github.com/bitcoin/bitcoin/pull/21832/commits/ed7feb460e841cec9570fc4115dc00486eec81b0) with the following changes:
1. Added space before `(₿)`: `s/Balance(₿):/Balance (₿):/` and `s/Balances(₿):/Balances (₿):/`
2. Combined Connections into `Network` header: `Network: in <in>, out <out>, total <total>`
3. Extended colors for `Chain: <chain>`, `Balances (₿)`, `Balance: (₿)`, `Network: in <in>, out <out>, total <total>`, `Wallet: <name>`
4. Right justify balances to the largest balance.

Single wallet 
![image](https://user-images.githubusercontent.com/49265907/118092111-3d831d00-b3fe-11eb-8a51-8f874b70ea98.png)
Multi wallet 
![image](https://user-images.githubusercontent.com/49265907/118092055-2c3a1080-b3fe-11eb-98ba-cfc0b69c1aeb.png)




> Oh, and if you need some sBTC (signet coins) for testing, here's a faucet: https://signetfaucet.com (I can send you some too, if you need).

I harder a fake wallet to simulate and test a large balance wallet and it should be working but if you don't mind sending me some sBTC that would be great(address `tb1q8rcmyqlphnpsemf8rgxpcwnlf0rvqnuqwhlzwf`). Thanks!

jonatack commented:
Sent you some signet coins (wow, the faucet used to be 10 sBTC and now it's max .01). Thanks for updating! Looking pretty good. I think the balances alignment should based on the decimal point (apologies, I was unclear) which might require padding spaces after the balance as well if there are fewer than 8 digits after the decimal point (I thought there were always 8). And maybe we can omit the `:` after the balances. WDYT? I'll re-test soon.

klementtan commented:
Got it! Thanks for sending it to me and replying to me so quickly. My apologies the amount was hardcoded by me and I added fewer decimal places than what it would actually have. Here is the updated screenshot of what it would look like when one wallet contains "longer" balance than others

![image](https://user-images.githubusercontent.com/49265907/118150344-071cc080-b445-11eb-9b8f-5de1f92a3a87.png)

> And maybe we can omit the `:` after the balances 
Yeah I agree `<balance> <wallet>` makes more sense as `<balance>` does not act as a key/fieldname. Will update it

jonatack commented:
Will re-test and review very soon.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +869,87 @@ static void GetWalletBalances(UniValue& result)
+        result_string += strprintf("Pay transaction fee: %s\n\n", result["paytxfee"].getValStr());
+    }
+    if (!result["balance"].isNull()) {
+        // \u20BF represents bitcoin emoji
```
This is clear from the `BITCOIN_EMOJI` constant, so I think you can drop this line.

klementtan replied:
Removed the comment

jonatack commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -47,53 +83,55 @@ def run_test(self):
         self.log.info("Test connecting with non-existing RPC cookie file")
-        assert_raises_process_error(1, "Could not locate RPC credentials", self.nodes[0].cli('-rpccookiefile=does-not-exist', '-rpcpassword=').echo)
+        assert_raises_process_error(1, "Could not locate RPC credentials",
+                                    self.nodes[0].cli('-rpccookiefile=does-not-exist', '-rpcpassword=').echo)
```
The diff in the above 15 lines in  the "Test -stdinrpcpass option" section is unrelated formatting changes. I checked that it doesn't change the code, but it's easier and safer to review without them and normally they shouldn't be added in this commit. I think they can probably be omitted.

klementtan replied:
My apologies for that 😅. Removed the unrelated formatting changes in this PR.

jonatack commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -25,6 +26,41 @@
+                value = ''
+            cli_get_info[key.strip()] = value.strip()
+        line_idx += 1
+    return cli_get_info
```
This seems like a lot of workaround, would it be easier to drop this and test the stdout instead? I think we do that in other tests (feature_help.py, tool_wallet.py, etc). I guess it could be done later.

jonatack reviewed (commented):
I think the result is pretty good now, thanks for working on the suggestions! A few code review comments below and then I think it might be ready.

ACK 3c866e5d79519f70fedd04e03683ab51e754e9ab modulo the comments below

jonatack commented:
ACK 6015d1d689c78dd5469cfde0963259422eef66df

![Screenshot from 2021-05-19 12-28-40](https://user-images.githubusercontent.com/2415484/118798538-473cd280-b88d-11eb-8125-55132ca4069b.png)

Two follow-ups:
- this may need a release note
- maybe simplify the tests per https://github.com/bitcoin/bitcoin/pull/21832#discussion_r635024249

jonatack commented:
(Also remove or update the screenshots in the PR description.)

One nice advantage of this change here is that `-getinfo` uses less vertical screen space.

klementtan commented:
@jonatack Thanks for taking the time to review the PR will work on remaining refactoring the test to simplify it.

jonatack commented:
> fixes: #17314

(This mention would auto-close the issue, I think. I don't know if it should be closed.)

kiminuo reviewed (commented):
Concept ACK

Adding screenshots to your [original post](https://github.com/bitcoin/bitcoin/pull/21832#issue-628579145) would be great.

Some thinking out loud: For me it's rather hard to say what keys are in `result` variable in `ParseGetInfoResult`. One way to solve that would be to introduce constants for strings like `version`, `blocks`, `headers`, `verificationprogress`, etc. so that an IDE can tell you where these JSON properties are used. This would help with refactorings too, I guess. The downside is that we would have more code. Feel free to ignore this note. If it resonates with you, it may be done in a follow-up PR. :)

jonatack commented:
> Some thinking out loud: For me it's rather hard to say what keys are in `result` variable in `ParseGetInfoResult`.

See the GetinfoRequestHandler class. I'm not sure the keys need to be hoisted out to a bunch of file-level static constants.

Maybe ParseGetInfoResult() could be a method in the GetinfoRequestHandler class for better encapsulatation.

klementtan commented:
Made the following changes:
- Updated s/`Time offset`/`Time offset (s)` and s/`Relay fee`/ `Relay fee (BTC/kB)` 
- Update release notes.
![image](https://user-images.githubusercontent.com/49265907/118904939-6ac75200-b94d-11eb-91ba-e99c8f965e1d.png)

> Maybe ParseGetInfoResult() could be a method in the GetinfoRequestHandler class for better encapsulation.

@jonatack I am not sure how to do that because `GetWalletBalances` takes in the return result of `GetinfoRequestHandler` and `ParseGetInfoResult` can only be called after `GetWalletBalances` is executed. Might need to refactor `GetWalletBalances` as well if we would like to encapsulate `ParseGetInfoResult`.

> Will work on refactoring the test to simplify it.

I branch off from this branch in my own repo to implement testing against stdout([PR here](https://github.com/klementtan/bitcoin/pull/1)) but I am not too sure if it actually simplifies the test because creating the expected stdout is quite complicated. @jonatack do you mind taking a look at it? If it is what you mean I would merge it into this branch and squash the commits. 

> Adding screenshots to your original post would be great.

@kiminuo My bad. I misunderstood previous comments and was under the impression that screenshots are not allowed in PR description. Updated OP to include PR description.

@kiminuo thanks for reviewing the PR

klementtan commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +869,86 @@ static void GetWalletBalances(UniValue& result)
+    result_string += strprintf("Version: %s\n", result["version"].getValStr());
+    result_string += strprintf("Time offset: %s\n", result["timeoffset"].getValStr());
+    result_string += strprintf("Proxy: %s \n", result["proxy"].getValStr());
+    result_string += strprintf("Relay fee: %s\n\n", result["relayfee"].getValStr());
```
Good suggestion! Updated to `BTC/kB` as per the unit in the [code](https://github.com/bitcoin/bitcoin/blob/master/src/rpc/net.cpp#L606)

klementtan commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +869,86 @@ static void GetWalletBalances(UniValue& result)
+        result["connections"]["total"].getValStr(),
+        RESET);
+    result_string += strprintf("Version: %s\n", result["version"].getValStr());
+    result_string += strprintf("Time offset: %s\n", result["timeoffset"].getValStr());
```
Updated `s`. Yup you're right it is in seconds([code](https://github.com/bitcoin/bitcoin/blob/master/src/rpc/net.cpp#L135))

kiminuo commented:
> * Updated s/`Time offset`/`Time offset (s)` and s/`Relay fee`/ `Relay fee (BTC/kB)`
>   ![image](https://user-images.githubusercontent.com/49265907/118904939-6ac75200-b94d-11eb-91ba-e99c8f965e1d.png)
> 
> * Update release notes.

Looks good. Thanks for the patch.

* "Pay transaction fee" should have a unit too, right?
* "Proxy" is empty. Maybe "Proxy: N/A" when there is no proxy? Or hide it altogether when there is no proxy? I have not checked possible values but it feels strage to show "Proxy: <and nothing here>".

Good work!

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +870,86 @@ static void GetWalletBalances(UniValue& result)
+
+    if (!result["has_wallet"].isNull()) {
+        const std::string walletname = result["walletname"].getValStr();
+        result_string += strprintf("%sWallet: %s%s\n", MAGENTA, walletname.empty() ? "\"\"" : walletname, RESET);
```
I think that instead of `""`, it is more common to use `[default wallet]`. 

At least these make me think so:

* https://github.com/bitcoin/bitcoin/blob/ea8b2e8e127b5a27424c1c08a1352ec1c9382c96/src/qt/bitcoingui.cpp#L401
* https://github.com/bitcoin/bitcoin/pull/13275/

jonatack replied:
In the CLI the default wallet has been `""`.

kiminuo replied:
Is that user-friendly? I mean `""` looks like an output for programmers and not for users. However, if it is a convention in CLI, then it's better to align with it.

____
One can possibly write also `default wallet ("")`

jonatack replied:
If you need to specify the default wallet in the CLI or RPC, you pass `""` or just nothing, e.g.:
```
$ bitcoin-cli -signet -rpcwallet="" -getinfo
$ bitcoin-cli -signet -rpcwallet=  -getinfo
```
If you pass something else, it won't be recognized as the default wallet (unless there is some way I'm not aware of)

kiminuo replied:
I see, that makes sense. Even though I was not suggesting to pass "default wallet" to a CLI command, just to make it clear that `""` means a default wallet in the text output. 

Anyway, all is good.

jonatack commented:
> @jonatack I am not sure how to do that because `GetWalletBalances` takes in the return result of `GetinfoRequestHandler` and `ParseGetInfoResult` can only be called after `GetWalletBalances` is executed. Might need to refactor `GetWalletBalances` as well if we would like to encapsulate `ParseGetInfoResult`.

Makes sense, and this can be in a follow-up.

jonatack commented on doc/release-notes.md:
```diff
@@ -141,6 +141,7 @@ Tools and Utilities
   useful to see if the node knows enough addresses in a network to use options
   like `-onlynet=<network>` or to upgrade to current and future Tor releases
   that support Tor v3 addresses only.  (#21595)
+- Update `-getinfo` to return data in user-friendly format that also reduces vertical space. (#21832)
```
Would be good to insert a blank line before it.

klementtan replied:
Added black line before it.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +870,86 @@ static void GetWalletBalances(UniValue& result)
+    result_string += strprintf("Version: %s\n", result["version"].getValStr());
+    result_string += strprintf("Time offset (s): %s\n", result["timeoffset"].getValStr());
+    result_string += strprintf("Proxy: %s \n", result["proxy"].getValStr());
+    result_string += strprintf("Relay fee (%\\/kB): %s\n\n", CURRENCY_UNIT, result["relayfee"].getValStr());
```
s/kB/kvB/

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +870,86 @@ static void GetWalletBalances(UniValue& result)
+    result_string += strprintf("Version: %s\n", result["version"].getValStr());
+    result_string += strprintf("Time offset (s): %s\n", result["timeoffset"].getValStr());
+    result_string += strprintf("Proxy: %s \n", result["proxy"].getValStr());
+    result_string += strprintf("Relay fee (%\\/kB): %s\n\n", CURRENCY_UNIT, result["relayfee"].getValStr());
```
s/kB/kvB/ (virtual size is used now, see #21752)

klementtan replied:
Changed the units to `kvB` for Relay fee and Pay transaction fee

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -18,6 +18,7 @@
 #include <util/system.h>
 #include <util/translation.h>
 #include <util/url.h>
+#include <policy/feerate.h>
```
sort

klementtan commented:
Update the PR with the following changes:
- If no `Proxy` then display `N/A` instead of an empty string
- Change all units to `vkB`

kiminuo commented:
This is my output on Windows 10 with colors turned for the Win32 platform:

## Windows Terminal running Powershell 7

![image](https://user-images.githubusercontent.com/58662979/118949209-470a0900-b959-11eb-8457-6d9a41184f6b.png)

(https://github.com/microsoft/terminal/ is not bundled with Windows by default and Powershell 7 is not bundled in Windows 10 by default too)

Note: Windows Terminal running PowerShell 5 works good too which only shows that Terminal is the thing that processes the colors.

## Command prompt (good old cmd)

![image](https://user-images.githubusercontent.com/58662979/118949408-7de01f00-b959-11eb-9bd7-d236b5b2c529.png)

(bundled with Windows by default)

Note that the same command run in Powershell (without Windows Terminal) show the same output.

## Going further

I'm not sure how to support colors on Windows properly in all scenarios but the following two pages gives useful intro:

* https://en.wikipedia.org/wiki/ANSI_escape_code#DOS,_OS/2,_and_Windows (summary)
* https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_special_characters?view=powershell-7.1#escape-e (colors in PowerShell)

Maybe somebody will pick this up in the future.

klementtan commented:
@kiminuo Thanks a lot for testing! Unfortunately, I do not own a Windows computer to test this myself. 

Does this mean the `WIN32` check [here](https://github.com/bitcoin/bitcoin/pull/21832/files#diff-134ccc4e5308f84fb9c03a9a1e1599209560370e49e30c7f545fa7b9ba7e07deR882-R896) is insufficient?

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +870,87 @@ static void GetWalletBalances(UniValue& result)
+        if (!result["unlocked_until"].isNull()) {
+            result_string += strprintf("Unlocked until: %s\n", result["unlocked_until"].getValStr());
+        }
+        result_string += strprintf("Pay transaction fee (%s/kvB): %s\n\n", CURRENCY_UNIT, result["paytxfee"].getValStr());
```
So that it is clear that this line refers to the `-paytxfee` configuration option without having to look in the code, and that it is actually a fee rate, maybe either s/Pay transaction fee/Paytxfee/ or
```suggestion
        result_string += strprintf("Transaction fee rate (-paytxfee) (%s/kvB): %s\n\n", CURRENCY_UNIT, result["paytxfee"].getValStr());
```
See also src/wallet/init.cpp:
```cpp
argsman.AddArg("-paytxfee=<amt>", strprintf("Fee rate (in %s/kvB) to add to transactions you send (default: %s)",
```

klementtan replied:
Updated to `Paytxfee` instead because `Transaction fee rate (-paytxfee)` seems a bit too long

jonatack commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -66,34 +105,30 @@ def run_test(self):
+            assert_equal(int(cli_get_info['Keypool size']), wallet_info['keypoolsize'])
+            assert_equal(int(cli_get_info['Unlocked until']), wallet_info['unlocked_until'])
+            assert_equal(Decimal(cli_get_info['Pay transaction fee (BTC/kvB)']), wallet_info['paytxfee'])
+            assert_equal(Decimal(cli_get_info['Relay fee (BTC/kvB)']), network_info['relayfee'])
```
```suggestion
            assert_equal(Decimal(cli_get_info['Min tx relay fee (BTC/kvB)']), network_info['relayfee'])
```
(see the getnetworkinfo help)

klementtan replied:
Agreed. Updated to `Relay fee` to `Min tx relay fee rate`

klementtan commented:
Updated the PR with the following changes: 
- Change `Relay fee` to `Min tx relay fee rate`
- Change `Pay transaction fee` to `Paytxfee`

Cleaned up OP to additional details on how to test as the PR comments have become rather long

jonatack commented:
Will re-test shortly. This would be nice for the upcoming release.

jonatack commented:
This is nicely improved after the feedback by @kiminuo.

Regarding the paytxfee and relay fee labels, I wonder if the longer versions wouldn't be more user-friendly and have the advantage of both clearly being fee rates (see screenshot), but I'm honestly not sure. Happy to re-ACK if you update.

![Screenshot from 2021-05-25 23-13-01](https://user-images.githubusercontent.com/2415484/119569497-54325800-bd9e-11eb-9f19-579aa0d3949c.png)

ACK 7c0cd204d48772b5f41c13197f25e7c8f7a24d99

prayank23 commented:
```js
Concept ACK
```

klementtan commented:
@jonatack Thanks for reviewing again! 
> Regarding the paytxfee and relay fee labels, I wonder if the longer versions wouldn't be more user-friendly and have the advantage of both clearly being fee rates (see screenshot)

Yup I agree. Having a longer version for `paytxfee` would allow for better standardization and user-friendliness. Have updated the PR to change `Paytxfee (BTC/kVb)` to `Transaction fee rate (-paytxfee) (BTC/kvB)`.

Screen shot new update:
![image](https://user-images.githubusercontent.com/49265907/119587209-c6d41f80-be00-11eb-9d81-a627fe7266de.png)

Can I also ask:
- Is it harder to review if I constantly squash my changes compared to squashing it at the end when everything is finalized? 
- Should I remove screenshots and *markdown toggles* from the PR description because it would be added as the commit message after the PR has been merged?

klementtan commented:
> but markdown toggles maybe yes to not rely on markdown for it to make sense

@jonatack Ok got it! Updated the PR description to remove the markdown toggles.

jonatack commented:
I think this is a nice improvement and it has an ACK by me and 3 Concept ACKs, if anyone else would like to review or test it.

prayank23 commented:
1. nit: Could have created a "topic specific branch" for PR as mentioned in [`CONTRIBUTING.md#contributor-workflow`](/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#contributor-workflow)

2. Compiled and tested on Ubuntu. Bitcoin symbol unicode character was missing. Maybe we can avoid using it. Either mention nothing as balance in Bitcoin Core wallet will only be for Bitcoin and nothing else or write `BTC` after amount.

![image](https://user-images.githubusercontent.com/13405205/120106535-ef4e7780-c17a-11eb-826e-adf2e02f4a0d.png)

After running `sudo apt install fonts-noto` and relaunch terminal

![image](https://user-images.githubusercontent.com/13405205/120106659-9a5f3100-c17b-11eb-942a-ae79a4b4d13d.png)

3. nit: Pink color for wallet looks weird. Will be better if we could use some other color similar or same as color used for 'balance'.

Will compile for Windows and test it on command prompt and PowerShell by tomorrow.

klementtan commented:
@prayank23 thanks a lot for reviewing ❤️ 

> 1. nit: Could have created a "topic specific branch" for PR as mentioned in [CONTRIBUTING.md#contributor-workflow](/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#contributor-workflow)

My apologies for that. I didn't want to create a new PR with a "topic-specific branch" as other people had already started reviewing the PR. 

> 2. Compiled and tested on Ubuntu. Bitcoin symbol unicode character was missing. Maybe we can avoid using it. Either mention nothing as balance in Bitcoin Core wallet will only be for Bitcoin and nothing else or write `BTC` after amount.

Yeah, that makes sense since the Bitcoin symbol isn't supported by default for all systems and it is just for novelty purposes. Removed the Bitcoin symbol in `ef535b4`

> 3. nit: Pink color for wallet looks weird. Will be better if we could use some other color similar or same as color used for 'balance'.

Hmmm I am not too sure about this as the only other color that is left in [ANSI Colour Code](https://en.wikipedia.org/wiki/ANSI_escape_code#3-bit_and_4-bit) is `red` and I am afraid that it would give a false sense of danger to the user. 

Another approach would be to use **bright** cyan(`\x1B[96m`) for `wallet`(refer to screenshot below). However in my opinion, the difference between **bright** and **non-bright** color is too subtle and having a distinct color difference(ie using `magenta`) would be better.

![image](https://user-images.githubusercontent.com/49265907/120112001-a44a5980-c1a6-11eb-882d-20ebfa3d8181.png)

klementtan commented:
Made the following changes for `a283d3d to ef535b4`:
- Removed bitcoin emoji as it is not supported by default on ubuntu 

<details>

<summary> Screenshots </summary>

*No wallet loaded:*
![image](https://user-images.githubusercontent.com/49265907/120110499-57fc1b00-c1a0-11eb-9f62-169e5031e569.png)

*Single wallet loaded*
![image](https://user-images.githubusercontent.com/49265907/120110548-7b26ca80-c1a0-11eb-9635-ae8c48a99206.png)

*Multi wallet loaded*
![image](https://user-images.githubusercontent.com/49265907/120110571-8974e680-c1a0-11eb-8c82-4ba10d707a6f.png)


</details>

prayank23 commented:
> Hmmm I am not too sure about this as the only other color that is left in ANSI Colour Code is red and I am afraid that it would give a false sense of danger to the user.

Agree. Red works better for errors.

> Another approach would be to use bright cyan(\x1B[96m) for wallet(refer to screenshot below). However in my opinion, the difference between bright and non-bright color is too subtle and having a distinct color difference(ie using magenta) would be better.

I am okay with this approach or pink is also fine if we don't have a better color.

prayank23 reviewed (approved):
tACK https://github.com/bitcoin/bitcoin/pull/21832/commits/ef535b48b618a40383498a35ec08a45992e51b58

klementtan commented:
[`ef535b4`](https://github.com/bitcoin/bitcoin/commit/ef535b48b618a40383498a35ec08a45992e51b58) to [`e94209b`](https://github.com/bitcoin/bitcoin/commit/e94209b4c5851d1d7aa2ae7b2af4a29b5123087f) rebase to resolve conflict.

kiminuo commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -68,34 +104,30 @@ def run_test(self):
             self.log.info("Test -getinfo and bitcoin-cli getwalletinfo return expected wallet info")
-            assert_equal(cli_get_info['balance'], BALANCE)
-            assert 'balances' not in cli_get_info.keys()
+            assert_equal(float(cli_get_info['Balance']), BALANCE)
```
nit: Comparing of floats in general can lead to comparing errors in general. (Please see https://stackoverflow.com/questions/4548004/how-to-correctly-and-standardly-compare-floats) 

So even though it may not be an issue now and here, it's still a question whether one should consider using `Decimal` comparison or even string comparison.

I don't know exact details here, so feel free to ignore this comment.

klementtan replied:
Yeah I agree, comparing floats directly is not a great idea and we could do something like [this](https://stackoverflow.com/a/33024979/11635160). However I think if we were to use that method a separate PR should be created to replace all float comparisons with it.

MarcoFalke replied:
The others in this file are already using Decimal, no? E.g. `assert_equal(Decimal(cli_get_info['Balance']), amounts[1])`.

klementtan replied:
Yup, they are all using python Decimal for the file. Correct if I am wrong but I think what @kiminuo meant is that using `assert_equal` for floats(`float`/`Decimal`) might lead to comparison errors and maybe another type of assertion should be used to check if the floats are "close enough".

However, I am and very new here, and might not know the nuance behind it.

klementtan replied:
Also changed `float(cli_get_info['Balance'])` to `Decimal(cli_get_info['Balance']` to standardize with the rest of the tests.

MarcoFalke replied:
`Decimal` is not a float

klementtan replied:
My apologies I was confusing myself. Please ignore whatever I said

jonatack reviewed (commented):
Tested ACK 23c8c5284acb18d8e960184e0b2d425c8a304004

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +870,86 @@ static void GetWalletBalances(UniValue& result)
+        result_string += strprintf("Transaction fee rate (-paytxfee) (%s/kvB): %s\n\n", CURRENCY_UNIT, result["paytxfee"].getValStr());
+    }
+    if (!result["balance"].isNull()) {
+        result_string += strprintf("%sBalance%s: %s\n\n", CYAN, RESET, result["balance"].getValStr());
```
micro-nit
```suggestion
        result_string += strprintf("%sBalance:%s %s\n\n", CYAN, RESET, result["balance"].getValStr());
```

klementtan replied:
👍  updated this in [`8233881`](https://github.com/bitcoin/bitcoin/commit/823388119bb1179e3667f86e82ab6817f87eb616)

klementtan commented:
[`23c8c52`](https://github.com/bitcoin/bitcoin/commit/23c8c5284acb18d8e960184e0b2d425c8a304004) to [`8233881`](https://github.com/bitcoin/bitcoin/commit/823388119bb1179e3667f86e82ab6817f87eb616)
* Added color to `:` for `Balance:` and `Warning:`
* Minor tests cleanup

jonatack commented on doc/release-notes.md:
```diff
@@ -145,6 +145,8 @@ Tools and Utilities
   like `-onlynet=<network>` or to upgrade to current and future Tor releases
   that support Tor v3 addresses only.  (#21595)
 
+- Update `-getinfo` to return data in user-friendly format that also reduces vertical space. (#21832)
```
```suggestion
- Update `-getinfo` to return data in a user-friendly format that also reduces vertical space. (#21832)
```

klementtan replied:
👍 Resolved it in [b634ff2](https://github.com/bitcoin/bitcoin/commit/b634ff26ab8cf4bdffed1a75300ec63a9da22cf1)

jonatack commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -27,6 +28,41 @@
 
+
+def cli_get_info_string_to_dict(cli_get_info_string):
+    # helper method to convert human readable -getinfo into a dictionary
```
```suggestion
    """Helper method to convert human readable -getinfo into a dictionary"""
```

klementtan replied:
👍 Resolved it in [`b634ff2`](https://github.com/bitcoin/bitcoin/commit/b634ff26ab8cf4bdffed1a75300ec63a9da22cf1)

jonatack commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -27,6 +28,41 @@
+        # Remove ansi colour code
+        line = ansi_escape.sub('', lines[line_idx])
+        if "Balances" in line:
+            # When Balances appears in a line, all of the following line contains "balance: wallet" until an empty line
```
```suggestion
            # When "Balances" appears in a line, all of the following lines contain "balance: wallet" until an empty line
```

klementtan replied:
Good catch! Resolved it in [`b634ff2`](https://github.com/bitcoin/bitcoin/commit/b634ff26ab8cf4bdffed1a75300ec63a9da22cf1)

jonatack reviewed (commented):
ACK 823388119bb1179e3667f86e82ab6817f87eb616

A few nits below, will re-ACK if you choose to address them.

klementtan commented:
> A few nits below, will re-ACK if you choose to address them.
@jonatack thanks! Addressed the nits in  [`8233881`](https://github.com/bitcoin/bitcoin/commit/8233881) to [`b634ff2`](https://github.com/bitcoin/bitcoin/commit/b634ff2)

kiminuo commented:
tACK b634ff2

Good work!

jonatack commented:
Code review re-ACK b634ff26ab8cf4bdffed1a75300ec63a9da22cf1 per `git diff 8233881 b634ff2`, documentation and code comment changes only

prayank23 commented:
re-ACK  https://github.com/bitcoin/bitcoin/commit/b634ff26ab8cf4bdffed1a75300ec63a9da22cf1

Although I couldn't' see any colors in PowerShell like https://github.com/bitcoin/bitcoin/pull/21832#issuecomment-844890803 :(

Maybe it works only with Microsoft Terminal.

### Command Prompt
<details><summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/13405205/120388647-0642c480-c349-11eb-989f-176695827444.png)
</details>

### PowerShell

<details><summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/13405205/120388559-e3181500-c348-11eb-9eae-81df8b10555b.png)
</details>

prayank23 commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +870,86 @@ static void GetWalletBalances(UniValue& result)
+{
+    if (!find_value(result, "error").isNull()) return;
+
+#ifndef WIN32
+    const std::string RESET = "\x1B[0m";
+    const std::string GREEN = "\x1B[32m";
+    const std::string BLUE = "\x1B[34m";
+    const std::string YELLOW = "\x1B[33m";
+    const std::string MAGENTA = "\x1B[35m";
+    const std::string CYAN = "\x1B[36m";
+#else
+    const std::string RESET = "";
+    const std::string GREEN = "";
+    const std::string BLUE = "";
+    const std::string YELLOW = "";
+    const std::string MAGENTA = "";
+    const std::string CYAN = "";
+#endif
```
```suggestion
const std::string RESET = "\x1B[0m";
const std::string GREEN = "\x1B[32m";
const std::string BLUE = "\x1B[34m"; 
const std::string YELLOW = "\x1B[33m";
const std::string MAGENTA = "\x1B[35m"; 
const std::string CYAN = "\x1B[36m";
```

If you remove the code that ignores ANSI escape sequence on Windows, we can see colors on Windows in Command Prompt and PowerShell as well

prayank23 reviewed (changes requested):
This is in Command Prompt after making the changes:

![image](https://user-images.githubusercontent.com/13405205/120402199-3a28e480-c35f-11eb-8241-70dd1b9190ef.png)

sipa commented:
Maybe this should only be enabled when stdout is connected to a terminal (`isatty(stdout)`)?

klementtan commented:
@prayank23 thanks a lot for testing!

> reg add HKEY_CURRENT_USER\Console /v VirtualTerminalLevel /t REG_DWORD /d 0x00000001 /f

If we go with that approach does it mean that bitcoin core will no longer work out of the box for Windows users? In my opinion not having colors for WIN32 would be better than having to run an additional step of entering the `reg add ...` command as an administrator. What do you think? :)

> @sipa: Maybe this should only be enabled when stdout is connected to a terminal (isatty(stdout))?

Is the purpose of this to not return ANSI codes when `-getinfo` is executed outside of the terminal? 

**Side note**: If the current handling of colored text is unsatisfactory I would prefer to remove all colored text support in this PR and create a separate PR for that. It seems that supporting colored text for every OS is quite complicated.

sipa commented:
@klementtan Right, when you're sending output to a text file or through grep or something like that, you generally don't want colorization.

prayank23 commented:
> If we go with that approach does it mean that bitcoin core will no longer work out of the box for Windows users? In my opinion not having colors for WIN32 would be better than having to run an additional step of entering the reg add ... command as an administrator. What do you think? :)

Colors don't work on Windows with ANSI escape sequence by default. So making the changes that I suggested in https://github.com/bitcoin/bitcoin/pull/21832#discussion_r643550352 will change nothing but simplify code as we will use same code for Linux and Windows.

If any user is curious and wants to use colors we can always suggest making changes in registry key. Else we will have to use `windows.h` and different code for Windows. IMO there are very less users who use Bitcoin Core on Windows and want to see colors in command prompt. So asking few power users who are curious to make changes in registry, use same code with less lines for every OS will be a better approach.

klementtan commented:
> Right, when you're sending output to a text file or through grep or something like that, you generally don't want colorization.

Good idea! Updated  b634ff2 to bc97f9a to return colored text when connected to the terminal.
<details>
<summary> Before  b634ff2 </summary>

```
./src/bitcoin-cli -signet -rpcwallet="" -getinfo > temp.txt
```
`temp.txt` contents
```
[34mChain: signet[0m
Blocks: 40452
Headers: 40452
Verification progress: 0.9999998746026817
Difficulty: 0.002836490385148087

[32mNetwork: in 0, out 10, total 10[0m
Version: 219900
Time offset (s): 0
Proxy: N/A
Min tx relay fee rate (BTC/kvB): 0.00001000

[35mWallet: ""[0m
Keypool size: 1000
Transaction fee rate (-paytxfee) (BTC/kvB): 0.00000000

[36mBalance:[0m 11.30000000

[33mWarnings:[0m This is a pre-release test build - use at your own risk - do not use for mining or merchant applications
```
</details>

<details>
<summary> After bc97f9a </summary>

```
./src/bitcoin-cli -signet -rpcwallet="" -getinfo > temp.txt
```
`temp.txt` contents
```
Chain: signet
Blocks: 40450
Headers: 40450
Verification progress: 0.9999923418449367
Difficulty: 0.002836490385148087

Network: in 0, out 10, total 10
Version: 219900
Time offset (s): 0
Proxy: N/A
Min tx relay fee rate (BTC/kvB): 0.00001000

Wallet: ""
Keypool size: 1000
Transaction fee rate (-paytxfee) (BTC/kvB): 0.00000000

Balance: 11.30000000

Warnings: This is a pre-release test build - use at your own risk - do not use for mining or merchant applications
```
</details>

prayank23 commented:
ACK https://github.com/bitcoin/bitcoin/commit/84547644885098fc94a3c126a0464f3d08277f5f

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +874,89 @@ static void GetWalletBalances(UniValue& result)
+    std::string BLUE = "";
+    std::string YELLOW = "";
+    std::string MAGENTA = "";
+    std::string CYAN = "";
```
style nit, std::string is default-initialized to `""`
```diff
-    std::string RESET = "";
-    std::string GREEN = "";
-    std::string BLUE = "";
-    std::string YELLOW = "";
-    std::string MAGENTA = "";
-    std::string CYAN = "";
+    std::string RESET, GREEN, BLUE, YELLOW, MAGENTA, CYAN;
```

klementtan replied:
Agreed that is much cleaner 👍 . Updated to this in 7f253cf

jonatack reviewed (commented):
ACK 84547644885098fc94a3c126a0464f3d08277f5f

Tested on debian that piping the output to a .txt file does not write the special characters corresponding to the color codes, whereas with the previous versions like b634ff26a it did.

kiminuo commented:
ACK 84547644885098fc94a3c126a0464f3d08277f5f

klementtan commented:
84547644885098fc94a3c126a0464f3d08277f5f to 7f253cf resolved style issues. No change in any behaviour to `-getinfo`.

jonatack commented:
re-ACK 7f253cf3459afa340412887765b62faff2f387f5 per `git diff 8454764 7f253cf`

prayank23 commented:
ACK https://github.com/bitcoin/bitcoin/commit/7f253cf3459afa340412887765b62faff2f387f5

lsilva01 commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +874,84 @@ static void GetWalletBalances(UniValue& result)
+    std::string result_string = strprintf("%sChain: %s%s\n", BLUE, result["chain"].getValStr(), RESET);
+    result_string += strprintf("Blocks: %s\n", result["blocks"].getValStr());
+    result_string += strprintf("Headers: %s\n", result["headers"].getValStr());
+    result_string += strprintf("Verification progress: %s\n", result["verificationprogress"].getValStr());
```
nit: instead of `verificationprogress: 0.9999...` or `verificationprogress: 0.9977977097643339`, this change shows `verificationprogress: 100%` and `verificationprogress: 99.78%` respectively 

```suggestion
    result_string += strprintf("Verification progress: %.2f%%\n", result["verificationprogress"].get_real() * 100);
```

klementtan replied:
Thanks a lot for reviewing! 

I actually considered changing it to 2 decimal places but was on the fence for it. My concern is that currently `0.01%` of the current height(`686181`) is about 6 blocks and displaying only 2 decimal places could confuse the user into thinking that their node is stuck. Additionally, as time goes on and the best height increases, I feel that it would be better to have a more granular value for verification progress.

What do you think? If the others also feel that having 2 decimal places is better I would change it! 😄

theStack replied:
I think Isilva01's proposal was primarily to change to per-cent unit, and the number of decimal places is just an example. Personally I'd also prefer showing in %; how about with 4 decimal places? With that, even a total block height of 1000000 (still years away) each single block verified results in a change of the verification progress display.

klementtan replied:
Yeah that makes sense! % would be better than fractional from. Implemented this in c5625c2. Thanks!

lsilva01 commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +874,84 @@ static void GetWalletBalances(UniValue& result)
+    result_string += strprintf("Difficulty: %s\n\n", result["difficulty"].getValStr());
+
+    result_string += strprintf(
+        "%sNetwork: in %s, out %s, total %s%s\n",
```
nit: Maybe `Connections` is clearer. `Network` can refer to IPv4, Tor, I2P ...

klementtan replied:
`Network` a "header" that is used to represent that the segment contains network-related information from `getnetworkinfo`. The connections information moved up to the same line as `Network` as their values are quite self-explanatory and we will save vertical real estate with that. https://github.com/bitcoin/bitcoin/pull/21832#issuecomment-840096163 for more information.

lsilva01 reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/21832/commits/7f253cf3459afa340412887765b62faff2f387f5 on Ubuntu 20.04.

luke-jr commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +874,84 @@ static void GetWalletBalances(UniValue& result)
+
+#ifndef WIN32
+    if (isatty(fileno(stdout))) {
+      // Only print colored text if OS is not WIN32 and stdout is connected to a terminal.
```
Might be nice to have a way to force it on/off?

klementtan replied:
Are you suggesting a `bitcon.conf` option or CLI argument to allow WIN32 users to print ANSI color code and non-WIN32 users to not print ANSI color code? For the former, I am planning on making a follow-up PR to allow that but I am not too sure if there is a strong use case for the latter tho.

luke-jr replied:
eg, most GNU tools support a `--color=always/never/auto`

Often useful when you want to pipe coloured output to `less` or such

klementtan replied:
Good suggestion! Added `-color` option in https://github.com/bitcoin/bitcoin/commit/5a2a6b3bc5e2b1bd085f5a687f1992182b71ad62

klementtan commented:
> @luke-jr: Might be nice to have a way to force it on/off?

7f253cf to 5a2a6b3: Added `-color=never/always/auto`
* `always`: Will always display color
* `never`: Will never display color
* `auto`: Will only display color if OS is not WIN32 and stdout connected to terminal 

### Tested scenarios 
<details>
<summary>-color not set</summary>

```bash
$ ./src/bitcoin-cli -signet -getinfo
$ ./src/bitcoin-cli -signet -getinfo | cat
```
![image](https://user-images.githubusercontent.com/49265907/121764376-f3649700-cb75-11eb-8394-8d5ffcdf987c.png)

</details>

<details>
<summary>-color=always</summary>

```bash
$ ./src/bitcoin-cli -signet -color=always -getinfo
$ ./src/bitcoin-cli -signet -color=always -getinfo | cat
```
![image](https://user-images.githubusercontent.com/49265907/121764514-1d6a8900-cb77-11eb-99cd-ee8fe8d0d8ea.png)
</details>

<details>
<summary>-color=never</summary>

```bash
$ ./src/bitcoin-cli -signet -color=never -getinfo
$ ./src/bitcoin-cli -signet -color=never -getinfo | cat
```
![image](https://user-images.githubusercontent.com/49265907/121764561-6d495000-cb77-11eb-87fa-3148c8a9e4a6.png)
</details>

<details>
<summary>-color=auto</summary>

```bash
$ ./src/bitcoin-cli -signet -color=auto -getinfo
$ ./src/bitcoin-cli -signet -color=auto -getinfo | cat
```

![image](https://user-images.githubusercontent.com/49265907/121764574-8d790f00-cb77-11eb-8176-7c790ae7d15a.png)

</details>

theStack commented:
Concept ACK
Only quickly tested the "No wallet" scenario so far, looks very nice!  🎉 
Planning to test with wallet(s) loaded and to code-review tomorrow.

klementtan commented:
> @theStack how about with 4 decimal places? 

5a2a6b3 to c5625c2: Chanted `Verification progress:` to percentage with 4 dp.

<details>
<summary> New screenshots </summary>

*No wallet loaded:*
![image](https://user-images.githubusercontent.com/49265907/121794631-94b62080-cc3c-11eb-8627-d0d1c25f0878.png)

*Single wallet loaded*
![image](https://user-images.githubusercontent.com/49265907/121794619-6df7ea00-cc3c-11eb-8420-d0c18236b188.png)

*Multi wallet loaded*
![image](https://user-images.githubusercontent.com/49265907/121794626-810aba00-cc3c-11eb-8cd7-c15ede1918ac.png)
</details>

theStack reviewed (approved):
Tested ACK c5625c2f8258b896a8dcf3d60646bf1f7bd0bbe5

Ran all three scenarios (no wallet, single wallet, multi-wallet) in OpenBSD 6.9 on mainnet and signet, works well and looks neat.
 
Now interactively watching the IBD progress via e.g.
```
watch -n 1 -d ./src/bitcoin-cli -getinfo
``` 
is quite a nice alternative to `tail -f ~/.bitcoin/debug.log` with less noise.

Some follow-up ideas I could imagine:
* also display date/time of latest block verified (maybe even the delta to the current time, to immediately see how far one is behind during IBD, or how long ago the last block was received, respectively)
* also show on-disk size of the block chain (`getblockchaininfo` returns a field `"size_on_disk"`)
* show whether pruning is enabled
* verification progress bar?

klementtan commented:
@theStack thanks for reviewing and testing. Yeah, I agree those are great ideas! I will update the issue with other improvement ideas after this PR gets merged. I would also like to refactor `ParseGetInfoResult` and `GetWalletBalances` into `GetinfoRequestHandler`.

achow101 commented:
Code Review ACK d6211e9a49fda31c2e2792874e537858c1913af3

klementtan commented:
c5625c2 to d6211e9 resolve typo in comments

theStack reviewed (approved):
re-ACK d6211e9a49fda31c2e2792874e537858c1913af3 🍸 
(the only change since my last ACK was fixing a debug message in the test)

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -77,6 +82,7 @@ static void SetupCliArgs(ArgsManager& argsman)
     argsman.AddArg("-stdin", "Read extra arguments from standard input, one per line until EOF/Ctrl-D (recommended for sensitive information such as passphrases). When combined with -stdinrpcpass, the first line from standard input is used for the RPC password.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-stdinrpcpass", "Read RPC password from standard input as a single line. When combined with -stdin, the first line from standard input is used for the RPC password. When combined with -stdinwalletpassphrase, -stdinrpcpass consumes the first line, and -stdinwalletpassphrase consumes the second.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-stdinwalletpassphrase", "Read wallet passphrase from standard input as a single line. When combined with -stdin, the first line from standard input is used for the wallet passphrase.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-color=<when>", "Enable key information from cli response to be displayed with color. With \"-color=auto\", bitcoin-cli will only emit color codes when standard output is connected to a terminal and OS is not WIN32. \"when\" can be \"never\", \"always\", or \"auto\"(default).", ArgsManager::ALLOW_STRING, OptionsCategory::OPTIONS);
```
Nit: I would expect:

```suggestion
    argsman.AddArg("-color=<when>", "Enable key information from CLI response to be displayed with color. With \"-color=auto\", bitcoin-cli will only emit color codes when standard output is connected to a terminal and OS is not WIN32. \"when\" can be \"never\", \"always\", or \"auto\"(default).", ArgsManager::ALLOW_STRING, OptionsCategory::OPTIONS);
```

klementtan replied:
👍  Changed in 14cb2e0

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +875,99 @@ static void GetWalletBalances(UniValue& result)
+
+    if (gArgs.IsArgSet("-color")) {
+        // Override default should_colorize value if -color is set.
+        std::string color = gArgs.GetArg("-color", "auto");
```
The idea is that when somebody passes `-color=nonsense`, then you consider the value to be `auto` (no warning or so), right?

klementtan replied:
Yup! If they don't enter a valid "when" value, it would be treated as the default colouring.

klementtan commented:
d6211e9 to 14cb2e0: to re-trigger CI because of failing flaky test and `s/cli/CLI` for args options.

Apologies to previous ACKs

prayank23 commented:
### 1. Terminal (-color not set):

<details><summary>Single Wallet</summary>

![image](https://user-images.githubusercontent.com/13405205/121998112-04591680-cdc9-11eb-999e-ea0fc8c5aac4.png)

</details>

<details><summary>Multi Wallet</summary>

![image](https://user-images.githubusercontent.com/13405205/121994709-6747af00-cdc3-11eb-9cde-0b52868d3e86.png)</details>

### 2. -color=never

```
bitcoin-cli -getinfo -color=never
bitcoin-cli -getinfo -color=never > color_never.txt
```
<details><summary>Terminal and Redirection</summary>

![image](https://user-images.githubusercontent.com/13405205/121997545-27cf9180-cdc8-11eb-9e09-de666c7fa1a9.png)

</details>

### 3. Redirect (-color not set and always):



<details><summary>-color not set</summary>

![image](https://user-images.githubusercontent.com/13405205/121996416-3157fa00-cdc6-11eb-82f8-8bb1cfb9ab5a.png)
</details>

```
bitcoin-cli -getinfo > color_notset.txt
```

<details><summary>-color=always</summary>

![image](https://user-images.githubusercontent.com/13405205/121996586-7f6cfd80-cdc6-11eb-875f-59db26c7a837.png)
</details>

```
bitcoin-cli -getinfo -color=always > color_always.txt
```

prayank23 reviewed (approved):
re-ACK https://github.com/bitcoin/bitcoin/pull/21832/commits/14cb2e0fe1384f89b4c9ca70751e6676f438b7a5

Tested again with new changes `-color` and redirection. LGTM.

theStack reviewed (approved):
re-ACK 14cb2e0fe1384f89b4c9ca70751e6676f438b7a5 🗻

achow101 commented:
re-ACK 14cb2e0fe1384f89b4c9ca70751e6676f438b7a5

lsilva01 reviewed (approved):
re-ACK https://github.com/bitcoin/bitcoin/pull/21832/commits/14cb2e0fe1384f89b4c9ca70751e6676f438b7a5

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -77,6 +82,7 @@ static void SetupCliArgs(ArgsManager& argsman)
     argsman.AddArg("-stdin", "Read extra arguments from standard input, one per line until EOF/Ctrl-D (recommended for sensitive information such as passphrases). When combined with -stdinrpcpass, the first line from standard input is used for the RPC password.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-stdinrpcpass", "Read RPC password from standard input as a single line. When combined with -stdin, the first line from standard input is used for the RPC password. When combined with -stdinwalletpassphrase, -stdinrpcpass consumes the first line, and -stdinwalletpassphrase consumes the second.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-stdinwalletpassphrase", "Read wallet passphrase from standard input as a single line. When combined with -stdin, the first line from standard input is used for the wallet passphrase.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-color=<when>", "Enable key information from CLI response to be displayed with color. With \"-color=auto\", bitcoin-cli will only emit color codes when standard output is connected to a terminal and OS is not WIN32. \"when\" can be \"never\", \"always\", or \"auto\"(default).", ArgsManager::ALLOW_STRING, OptionsCategory::OPTIONS);
```
nit, sort and missing space (see proposed improvement diff)
```suggestion
    argsman.AddArg("-color=<when>", "Enable key information from CLI response to be displayed with color. With \"-color=auto\", bitcoin-cli will only emit color codes when standard output is connected to a terminal and OS is not WIN32. \"when\" can be \"never\", \"always\", or \"auto\" (default).", ArgsManager::ALLOW_STRING, OptionsCategory::OPTIONS);
```

klementtan replied:
Resolved this in 658ab9c. Thanks!

jonatack reviewed (commented):
Lightly-tested re-ACK 14cb2e0fe1384f89b4c9ca70751e6676f438b7a5 with suggested changes to the new `-color` help (feel free to ignore the code suggestions)

<details><summary>diff</summary><p>

```diff
@@ -52,6 +52,9 @@ static constexpr int8_t UNKNOWN_NETWORK{-1};
 /** Default number of blocks to generate for RPC generatetoaddress. */
 static const std::string DEFAULT_NBLOCKS = "1";
 
+/** Default -color setting. */
+static const std::string DEFAULT_COLOR_SETTING{"auto"};
+
 static void SetupCliArgs(ArgsManager& argsman)
 {
     SetupHelpOptions(argsman);
@@ -70,6 +73,7 @@ static void SetupCliArgs(ArgsManager& argsman)
     argsman.AddArg("-netinfo", "Get network peer connection information from the remote server. An optional integer argument from 0 to 4 can be passed for different peers listings (default: 0). Pass \"help\" for detailed help documentation.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
 
     SetupChainParamsBaseOptions(argsman);
+    argsman.AddArg("-color=<when>", strprintf("Color setting for CLI output (default: %s). Valid values: always, auto (add color codes when standard output is connected to a terminal and OS is not WIN32), never.", DEFAULT_COLOR_SETTING), ArgsManager::ALLOW_STRING, OptionsCategory::OPTIONS);
     argsman.AddArg("-named", strprintf("Pass named instead of positional arguments (default: %s)", DEFAULT_NAMED), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-rpcclienttimeout=<n>", strprintf("Timeout in seconds during HTTP requests, or 0 for no timeout. (default: %d)", DEFAULT_HTTP_CLIENT_TIMEOUT), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-rpcconnect=<ip>", strprintf("Send commands to node running on <ip> (default: %s)", DEFAULT_RPCCONNECT), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
@@ -82,7 +86,6 @@ static void SetupCliArgs(ArgsManager& argsman)
     argsman.AddArg("-stdin", "Read extra arguments from standard input, one per line until EOF/Ctrl-D (recommended for sensitive information such as passphrases). When combined with -stdinrpcpass, the first line from standard input is used for the RPC password.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-stdinrpcpass", "Read RPC password from standard input as a single line. When combined with -stdin, the first line from standard input is used for the RPC password. When combined with -stdinwalletpassphrase, -stdinrpcpass consumes the first line, and -stdinwalletpassphrase consumes the second.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-stdinwalletpassphrase", "Read wallet passphrase from standard input as a single line. When combined with -stdin, the first line from standard input is used for the wallet passphrase.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-color=<when>", "Enable key information from CLI response to be displayed with color. With \"-color=auto\", bitcoin-cli will only emit color codes when standard output is connected to a terminal and OS is not WIN32. \"when\" can be \"never\", \"always\", or \"auto\"(default).", ArgsManager::ALLOW_STRING, OptionsCategory::OPTIONS);
 }
 
 /** libevent event log callback */
@@ -895,8 +898,7 @@ static void ParseGetInfoResult(UniValue& result)
 #endif
 
     if (gArgs.IsArgSet("-color")) {
-        // Override default should_colorize value if -color is set.
-        std::string color = gArgs.GetArg("-color", "auto");
+        const std::string color{gArgs.GetArg("-color", DEFAULT_COLOR_SETTING)};
         if (color == "always") {
             should_colorize = true;
         } else if (color == "never") {
```
</p></details>

klementtan commented:
14cb2e0 to 658ab9c: 

* Resolved stylying nit
* Add validation when invalid value added to `-color`

```
$ ./src/bitcoin-cli -signet -color=foo -getinfo
error: Invalid value for -color option. Valid values: always, auto, never.
```

jonatack reviewed (commented):
ACK 658ab9caa3423aa7f50ae7bd0ee7ebb50f0114b0 per `git diff 14cb2e0 658ab9c` and lightly re-tested rebased to master on each chain and with various wallet configs.

New -color option help
```
$ ./src/bitcoin-cli -h | grep -A1 color
  -color=<when>
       Color setting for CLI output (default: auto). Valid values: always, auto
       (add color codes when standard output is connected to a terminal
       and OS is not WIN32), never.
```

New error
```
$ ./src/bitcoin-cli -regtest -color=foo -getinfo
error: Invalid value for -color option. Valid values: always, auto, never.
```

klementtan commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -65,37 +101,43 @@ def run_test(self):
         self.log.info("Test -getinfo with arguments fails")
         assert_raises_process_error(1, "-getinfo takes no arguments", self.nodes[0].cli('-getinfo').help)
 
+        self.log.info("Test -getinfo with -color=never retuns no ANSI escape codes")
```
Good catch! Resolved this in d525177

klementtan commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -65,37 +101,43 @@ def run_test(self):
+        self.log.info("Test -getinfo with -color=never retuns no ANSI escape codes")
+        assert "\u001b[0m" not in self.nodes[0].cli('-getinfo', '-color=never').send_cli()
+
+        self.log.info("Test -getinfo with -color=always retuns ANSI escape codes")
```
Good catch! Resolved this in d525177

jonatack commented:
re-ACK d525177aba60e996481007989422280f9c191f15 per `git diff 658ab9c d525177`

Only change is addressing feedback on 2 test logging touch-ups.

theStack commented on src/bitcoin-cli.cpp:
```diff
@@ -867,6 +878,100 @@ static void GetWalletBalances(UniValue& result)
+            should_colorize = true;
+        } else if (color == "never") {
+            should_colorize = false;
+        } else if (color != DEFAULT_COLOR_SETTING) {
```
nit: I don't think using the constant is a good idea here. If the default color setting is ever changed, then passing "auto" would not work anymore.
```suggestion
        } else if (color != "auto") {
```

klementtan replied:
Good point resolved this in 62aaff1.

klementtan commented:
658ab9c to 62aaff1: resolved comments from reviewers regarding logging and better handling of `-color` option 🙏

theStack reviewed (approved):
re-ACK 62aaff178a665d757b604ee4a90556242c1135d3 🍪 

Verified that the following changes since my previous ACK are OK: stricter `-color` argument handling with a corresponding test, plus minor style changes in the `-color` help and the test log messages 👌

jonatack commented:
re-ACK 62aaff178a665d757b604ee4a90556242c1135d3 per `git diff d525177 62aaff1`

MarcoFalke commented on doc/release-notes.md:
```diff
@@ -145,6 +145,8 @@ Tools and Utilities
   that support Tor v3 addresses only.  (#21595)
 
+- Update `-getinfo` to return data in a user-friendly format that also reduces vertical space. (#21832)
+
```
Note to myself: Might have been better to merge before #22515.

maflcko reviewed (commented):
Needs rebase

theStack reviewed (approved):
re-ACK a37e29d32fde8c7b4143322deeef2a8a06114d43

Verified via `git range-diff 62aaff17...a37e29d3` that the latest change since my last ACK is only rebase-related.

jonatack commented:
Congrats @klementtan, well-deserved.

theStack commented:
This PR seems to be merged, but it still open. Another GitHub problem I guess? 🤔
